src/aws-api-mcp-server/awslabs/aws_api_mcp_server/core/aws/service.py

```diff
@@ -0,0 +1,227 @@
+        else f'{AWS_API_MCP_PROFILE_NAME} profile'
+    )
+    logger.info(
+        f'AWS credentials successfully resolved using: {creds_source}.'
+    )  # logging the source of credentials
+
+    return Credentials(
+        access_key_id=aws_creds.access_key,
+        secret_access_key=aws_creds.secret_key,
+        session_token=aws_creds.token,
+    )
+
+
+def is_operation_read_only(ir: IRTranslation, read_only_operations: ReadOnlyOperations):
+    """Check if the operation in the IR is read-only."""
+    if (
+        not ir.command_metadata
+        or not getattr(ir.command_metadata, 'service_sdk_name', None)
+        or not getattr(ir.command_metadata, 'operation_sdk_name', None)
+    ):
+        raise RuntimeError(
+            "failed to check if operation is allowed: translated command doesn't include service and operation name"
+        )
+
+    service_name = ir.command_metadata.service_sdk_name
+    operation_name = ir.command_metadata.operation_sdk_name
+    return read_only_operations.has(service=service_name, operation=operation_name)
+
+
+def validate(ir: IRTranslation) -> ProgramValidationResponse:
+    """Translate the given CLI command and return a validation response."""
+    classification = (
+        ProgramClassificationMetadata(**ir.classification.as_metadata())
+        if ir.classification
+        else None
+    )
+
+    return ProgramValidationResponse(
+        missing_context_failures=_to_missing_context_failures(ir.missing_context_failures),
+        validation_failures=_to_validation_failures(ir.validation_or_translation_failures),
+        classification=classification,
+    )
+
+
+def execute_awscli_customization(
+    cli_command: str,
+) -> AwsCliAliasResponse | AwsApiMcpServerErrorResponse:
+    """Execute the given AWS CLI command."""
+    args = split_cli_command(cli_command)[1:]
+
+    try:
+        stdout_capture = StringIO()
+        stderr_capture = StringIO()
+
+        with (
+            contextlib.redirect_stdout(stdout_capture),
+            contextlib.redirect_stderr(stderr_capture),
+        ):
+            driver.main(args)
+
+        stdout_output = stdout_capture.getvalue()
+        stderr_output = stderr_capture.getvalue()
+
+        return AwsCliAliasResponse(response=stdout_output, error=stderr_output)
+    except Exception as e:
+        return AwsApiMcpServerErrorResponse(
+            error=True,
+            detail=f"Error while executing '{cli_command}': {e}",
+        )
+
+
+def interpret_command(
+    cli_command: str,
+    credentials: Credentials,
+    default_region: str,
+    max_results: int | None = None,
+) -> ProgramInterpretationResponse:
+    """Interpret the given CLI command and return an interpretation response."""
+    interpreted_program = _interpret_command(
+        cli_command,
+        access_key_id=credentials.access_key_id,
+        secret_access_key=credentials.secret_access_key,
+        session_token=credentials.session_token,
```

Comment:
Do profiles work without the credentials being sent?

Reply:
Boto3 resolves credentials automatically when profile is used as the cred source, so we should have them at this point.